### ISSUE TITLE
Documentation issue: StiPreload causes circular dependency issues

### ISSUE THREAD

**Issue by marvinthepa** [rails/rails#45307]
### tl;dr

The way how to autoload as described in the  ["Single Table Inheritance" section in the "Autoloading and Reloading Constants" guide](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#single-table-inheritance) seems to be broken with Rails 7.

### Steps to reproduce

Check out https://github.com/marvinthepa/sti_preload_broken. Follow the steps described in https://github.com/marvinthepa/sti_preload_broken/blob/main/README.md

This repo is a very minimal rails application that uses STI and the recommended way to load the STI hierarchy as described in https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#single-table-inheritance

### Expected behavior

When loading the `Dog` STI model from the controller, the `Animal` parent model should load without error.

### Actual behavior

An error is thrown.
```
uninitialized constant Dog
```

Breakdown on why this happens:

1. The controller references the Dog model, which zeitwerk tries to load.
2. As `Dog < Animal`, zeitwerk tries to load `Animal`
3. Animal contains a `before_save` hook, which is installed, eventually leading to `ActiveSupport::Callbacks::__update_callbacks` being called.
4. As of ffae3bd8d69, `__update_callbacks` calls `Animal.descendants`
5. StiPreload kicks in, and finds and entry with `type == Dog` in the db, and tries to load it
6. As zeitwerk is already in the process of loading `Dog`, it fails the above mentioned error message

I have only found a very stupid workaround, which I do not want to keep using:

```ruby
      def descendants
        # HACK
        return super unless caller.grep(/__update_callbacks/).empty?

        preload_sti unless preloaded
        super   
      end
```

Maybe there is some way to ask zeitwerk which classes it currently tries to load and skip loading those?
Anyway, I think the documentation should be updated with a `StiPreload` that also works with rails 7.

### System configuration
**Rails version**: 7.0.3

**Ruby version**: 3.0.2p107

**Comment by p8** [comment]
Hey @marvinthepa,
Does the STI section in the Autoloading guide help?
https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#single-table-inheritance

**Comment by marvinthepa** [comment]
cc @fxn: as you wrote the documentation originally, do you have any ideas on how to fix this?

**Comment by fxn** [comment]
@marvinthepa We need to do something with that section.

Off the top of my head, you could move the same idea to a `to_prepare` block (untested):

```ruby
# config/initializers/eager_load_stis.rb
unless Rails.configuration.eager_load
  Rails.configuration.to_prepare do
    [Animal, Shape].each do |sti_root|
    # perform the query + constantize on sti_root
  end
end
```

The difference is that this is not lazy, you always eager load, but at least it works (it it does :)).

**Comment by wollistik** [comment]
@marvinthepa Ran into the same issue when upgrading our application. I played around with other options, like using the Zeitwerk `on_load` hook, but nothing did solve the issue. The STI base classes have to be explicitly made available before any subclass implicitly does in order to not running into issues.

@fxn your eager loading approach is currently the only way to reliable achieve this - i tested it and it works :+1: May be this should also be the official documented approach since others will fall into this issue as well.

BTW: There are always some gotchas with STI - see this approach in [Rails 5](https://guides.rubyonrails.org/v5.0/autoloading_and_reloading_constants.html#autoloading-and-sti). I think one simply have to live with it :wink:

**Comment by jrochkind** [comment]
I discovered this myself and reported as an issue at https://github.com/rails/rails/issues/46625

There are links to a simple reproduction app there should anyone want one. 

The Rails Guide is definitely telling people to do something that will result in mystery errors, so a fix would be nice. 

It was helpful to find this here; I am trying the initializer approach, with `to_prepare`. it seems to be working okay. Glad I found this, as before I was trying without `to_prepare`, which I realize would have been a problem (not doing the preload after reload) -- this stuff is a bit tricky to get right, so guide solution would be great. 

Yes, there seems to be no way around have to explicitily mention all your STI base classes  -- or, all descendant classes if you want to skip the trip to the DB -- I guess the Rails 5 guide solution wollistik linked to required explicitly listing all leaf classes with the (now no longer available) `require_dependency`, so this has been with us for a while -- Rails 6 and 6.1 briefly gave us a way out where all classes didn't have to be listed, but I guess that's gone to us again in Rails 7. 

(It's still a bit of a mystery to me why this worked okay in Rails 6/6.1... somehow it handled the circular dependency load there, or at least it didn't cause a raise?)

**Comment by jrochkind** [comment]
OK discovered a problem with y'alls recommended `to_prepare` in an initializer. 

If you do a `db:drop ; db:create` in your app (say in dev or test), the initializer to_prepare block for me is called too early when I try `rails db:create`.  It is called before the DB exists, and then in my implementation it tries to go to the DB to fetch class names, which errors becuase there is no DB yet... resulting in a really confusing error message from the `rails db:create` saying basically "we could not create your database because the database doesn't exist". 

Any suggestions for how to get it in `to_prepare`, but not when the db doesn't exist yet? I guess I could just put a rescue in. That seems horrible.

**Comment by fxn** [comment]
The `StiPreload` module has proven to be problematic. Also, it is a verbose solution that you need to copy & paste and depends on Active Record internals. It was an interesting idea, but I believe we need to retire it until we find a robust and transparent solution that ships in Active Record itself.

Until that moment arrives, I have [rewritten the documentation](https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html) to cover 4 ways to address STIs in a lazy load setting.

These options don't mess with `descendants`, it is all easy to understand public API.

In some cases you need to have full control on load order. For example, consider:

```ruby
class Base < ApplicationRecord
  DEFAULT = Child
end

class Child < Base
end
```

There, you cannot load `Child` first. You need to force `Base` to be loaded first.

**Comment by jrochkind** [comment]
Thanks @fxn !

[Option #4: Preload Types from the Database](https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#single-table-inheritance), involving  is the most generic and most similar to previous `StiPreload`. Thanks for including it. 

However, as written it can cause problems for your app, as it makes your app un-bootable without a database. 

You can reproduce by simply running:

    rails db:drop
    rails db:prepare

A more common place you might run into this is in CI -- if you build a fresh VM/container for each CI, and deploy your code with Option 4 in it, you are unable to run `rails db:prepare`, or any other rake/rails tasks at all. 

Here is one really hacky workaround to that that I came up with:

```ruby
unless Rails.application.config.eager_load
  Rails.application.config.to_prepare do
    types = Shape.unscoped.select(:type).distinct.pluck(:type)
    types.compact.each(&:constantize)
  rescue ActiveRecord::NoDatabaseError, ActiveRecord::StatementInvalid => e
    Rails.logger.debug("Could not pre-load Shape Single-Table Inheritance descendents: #{e.inspect}")
  end
end
```

I don't love this solution. The code is starting to look convoluted and hard to read, and I don't love that it might be swallowing other exceptions caused by things other than the database not existing.  (`StatementInvalid` is in there because it's what I got if the DB itself existed, but without the schema having been loaded yet; but of course there are many other things that could cause that exception too). 

I am curious if anyone has any other suggestions. 

I wonder if Option #4 as it is is not complete enough to include in the Guide, as it will probably cause people problems as above.

**Comment by fxn** [comment]
Oh sure, I was implicitly assuming that if you have code that hits the database when the application boots, for this, or for doing any other kind of preflight check, you are assuming the database exists.

Because if you need the database to boot, I cannot imagine a situation in which the application should be robust to not having a database created. You get an error and say, damn, there is no database, of course I cannot verify if there is an admin user seeded. But the application is unusable anyway, I do need to create the database.

Happy to tweak the example, or add a note about this.

**Comment by jrochkind** [comment]
Hm, I'm not sure if I am not explaining it clearly, or if you are just not seeing it the same way. 

In a typical rails application  you can run `rails db:prepare` if you have a new checkout with no DB yet existing/setup. 

Once you have added the `to_prepare` block as recommended in Option 4 in the guide, you can _no longer_ run `rails db:prepare` on a fresh checkout without a DB.  So it breaks that use case. One place that use case is common in the actually existing world is in CI.  but it can also apply to setting up a new staging server, etc. 

So when you say "I cannot imagine a situation in whcih the application should be robust to not having a database created", does it help when I explain that situation? It is in fact a very common situation. I agree that one does not expect a Rails application that has a database to be able to serve requests without the database existing. But one does expect to be able to run rake/rails tasks -- such as the tasks that _create_ a database!  

I am not sure how the example is best tweaked, and would love to hear ideas on it. I am not totally happy with what did above.

**Comment by fxn** [comment]
Hey, just a heads up, I'll come back to this thread.

**Comment by helmutrs** [comment]
@fxn  Thanks, I hope we can find a solution to this, I cannot upgrade to Rails 7 because of STI,  if I come up with an idea at least, I will post it as well. cheers.

**Comment by fxn** [comment]
@helmutrs Please don't get blocked by this. My intention is to delete that technique from the docs for several reasons, but need to ponder it. To unblock, you could for example implement some of the other patters, and as a last resort preload all constants one by one if you need fine control on order of evaluation (because in some scenarios you cannot just autoload as you walk a directory or database table, you might need to force certain order that respects dependencies).

**Comment by feliperaul** [comment]
@fxn The updated edge guides are very helpful, thank you so much! The preload directory (collapsed or not) was very helpful.

If you don't mind me asking a related question, I often find myself wanting to have a list of descendants class of a base class, and I'd like your input on how you approach these scenarios.

Imagine some e-mail template generators that all inherit from`< EmailCampaigns::Generators::Base`. 

Calling `EmailCampaigns::Generators::Base.descendants` will return an empty or incomplete array unless all the descendent classes are all preloaded.

However, the app is growing bigger and I'd like to avoid having to resort to the _preload directory_ to not harm performance in development (specially because our app seems to be having this need quite often). 

I generally don't mind in manually declaring the descendants classes beforehand in some kind of 'registry' pattern, but then the registry can't be in `::Base` otherwise the circular dependency you mentioned here arrives:

```
class EmailCampaigns::Generators::Base
  GENERATORS = [Child]
end

class EmailCampaigns::Generators::Child < EmailCampaigns::Generators::Base
end
```

So what I end up doing is defining a `email_campaigns/generators.rb` module and using that file to define the manual registry.

How do you usually approach this need for a *complete* list of descendants classes? Do you just eager load them all with the _preload directory_ like in the guide, or do you also manually register the descendants beforehand, or some third way I'm not seeing?

**Comment by fxn** [comment]
@feliperaul in general, my approach is to eager load what has to be eager loaded because that is what the semantics are aligned with.

So, while one may make an effort to make applications be as lean as possible when they boot, at some point you need to consider something as required during boot, and I do for these use cases.

Also, if the app is big and you run

```
bin/rails runner 'p $LOADED_FEATURES.size'
```

probably that will be several thousands. Do 10 files more matter? Probably not. So you do those 10 more, and live in peace because the STIs are loaded, the descendants are loaded, the APIs and semantics have no surprises, and there is nothing to mantain manually.

Having said that :), if you are still willing to mantain a registry, I believe a list of class names to be lazily constantized would work. Something like

```ruby
class MyBase
  DESCENDANTS = %w(...)
  private_constant :DESCENDANTS

  def self.descendants
    @descendants ||= DESCENDANTS.map { _1.constantize }
  end
end
```

I wrote that off the top of my head, but I believe it should be OK.

**Comment by feliperaul** [comment]
@fxn Thank you *very much* for that thorough explanation!
